File: tests/test_telemetry.c
#include "telemetry.h"
#include <zephyr/drivers/sensor.h>
#include <hal/nrf_adc.h>
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>

struct device fake_sensor;
struct adc adc;
static int64_t now;
static int temp_failure, adc_failure;
static struct sensor_value temperature;
int device_is_ready(const struct device *dev) { (void)dev; return 1; }
int sensor_sample_fetch(const struct device *dev) { (void)dev; return temp_failure; }
int sensor_channel_get(const struct device *dev, int channel, struct sensor_value *value)
{ (void)dev; (void)channel; *value = temperature; return 0; }
int64_t k_uptime_get(void) { return now; }
void k_msleep(int ms)
{
    now += ms;
    if (!adc_failure) { adc.EVENTS_END = 1; }
}
static void sample(int degrees, int fraction, unsigned raw, unsigned encoded_temp, unsigned voltage)
{
    uint8_t msd[16];
    memset(msd, 0xaf, sizeof msd);
    temperature = (struct sensor_value){degrees, fraction};
    adc.RESULT = raw;
    adc.TASKS_STOP = 0;
    assert(!od_read_msd(msd));
    for (unsigned i = 0; i < 13; i++) { assert(msd[i] == 0xaf); }
    assert(msd[13] == encoded_temp && msd[14] == (voltage & 255));
    assert(msd[15] == (0xae | (voltage >> 8)));
    assert(adc.CONFIG == 14 && adc.TASKS_STOP && !adc.ENABLE);
}
int main(void)
{
    sample(25, 250000, 853, 130, 300);
    sample(-10, -750000, 512, 59, 180);
    sample(-50, 0, 0, 0, 0);
    sample(100, 0, 1023, 255, 360);
    uint8_t msd[16], previous[16];
    memset(msd, 0xff, sizeof msd);
    memcpy(previous, msd, sizeof msd);
    adc_failure = 1;
    int64_t started = now;
    assert(od_read_msd(msd) == -ETIMEDOUT && now - started == 50);
    assert(!memcmp(previous, msd, sizeof msd) && !adc.ENABLE && adc.TASKS_STOP);
    temp_failure = 1;
    adc.TASKS_START = 0;
    assert(od_read_msd(msd) == -EIO && !adc.TASKS_START);
    assert(!memcmp(previous, msd, sizeof msd));
    puts("Telemetry encoding, snapshot preservation and ADC timeout/shutdown passed");
}

File: scripts/test_telemetry.py
"""Exercise the production sampler's encoding, atomic snapshot and ADC shutdown."""
from pathlib import Path
import os
import subprocess
import tempfile

root = Path(__file__).resolve().parents[1]
headers = {
    "zephyr/drivers/sensor.h": """
#include <stdint.h>
struct device { int unused; };
struct sensor_value { int32_t val1, val2; };
extern struct device fake_sensor;
#define DEVICE_DT_GET_ONE(x) (&fake_sensor)
#define SENSOR_CHAN_DIE_TEMP 0
int device_is_ready(const struct device *dev);
int sensor_sample_fetch(const struct device *dev);
int sensor_channel_get(const struct device *dev, int channel, struct sensor_value *value);
""",
    "zephyr/kernel.h": """
#include <stdbool.h>
#include <stdint.h>
#define CLAMP(x, lo, hi) ((x) < (lo) ? (lo) : (x) > (hi) ? (hi) : (x))
int64_t k_uptime_get(void);
void k_msleep(int ms);
""",
    "hal/nrf_adc.h": """
#include <stdint.h>
struct adc { uint32_t ENABLE, CONFIG, EVENTS_END, TASKS_START, TASKS_STOP, RESULT; };
extern struct adc adc;
#define NRF_ADC (&adc)
#define ADC_CONFIG_RES_10bit 2
#define ADC_CONFIG_RES_Pos 0
#define ADC_CONFIG_INPSEL_SupplyOneThirdPrescaling 3
#define ADC_CONFIG_INPSEL_Pos 2
""",
}
with tempfile.TemporaryDirectory(prefix="lt213a-telemetry-") as directory:
    temporary = Path(directory)
    for name, content in headers.items():
        path = temporary / name
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(content)
    binary = temporary / "test-telemetry"
    subprocess.run([os.environ.get("CC", "cc"), "-std=c11", "-Wall", "-Wextra", "-Werror",
        "-fsanitize=address,undefined", "-Isrc", f"-I{temporary}", "src/telemetry.c",
        "tests/test_telemetry.c", "-o", str(binary)], cwd=root, check=True)
    subprocess.run([str(binary)], check=True)
